cmd/thanos/sidecar.go

```diff
@@ -44,100 +43,53 @@ import (
-
-	dataDir := cmd.Flag("tsdb.path", "Data directory of TSDB.").
-		Default("./data").String()
-
-	reloaderCfgFile := cmd.Flag("reloader.config-file", "Config file watched by the reloader.").
-		Default("").String()
-
-	reloaderCfgOutputFile := cmd.Flag("reloader.config-envsubst-file", "Output file for environment variable substituted config file.").
-		Default("").String()
-
-	reloaderRuleDirs := cmd.Flag("reloader.rule-dir", "Rule directories for the reloader to refresh (repeated field).").Strings()
-
-	objStoreConfig := regCommonObjStoreFlags(cmd, "", false)
-
-	uploadCompacted := cmd.Flag("shipper.upload-compacted", "If true sidecar will try to upload compacted blocks as well. Useful for migration purposes. Works only if compaction is disabled on Prometheus. Do it once and then disable the flag when done.").Default("false").Bool()
-
-	ignoreBlockSize := cmd.Flag("shipper.ignore-unequal-block-size", "If true sidecar will not require prometheus min and max block size flags to be set to the same value. Only use this if you want to keep long retention and compaction enabled on your Prometheus instance, as in the worst case it can result in ~2h data loss for your Thanos bucket storage.").Default("false").Hidden().Bool()
-
 	minTime := thanosmodel.TimeOrDuration(cmd.Flag("min-time", "Start of time range limit to serve. Thanos sidecar will serve only metrics, which happened later than this value. Option can be a constant time in RFC3339 format or time duration relative to current time, such as -1d or 2h45m. Valid duration units are ms, s, m, h, d, w, y.").
 		Default("0000-01-01T00:00:00Z"))
 
 	m[component.Sidecar.String()] = func(g *run.Group, logger log.Logger, reg *prometheus.Registry, tracer opentracing.Tracer, _ <-chan struct{}, _ bool) error {
-		rl := reloader.New(
+
+		conf := sidecarComponentConfig{
+			grpc:       parseFlagsToGrpcConfig(cmd),
+			http:       parseFlagsToHttpConfig(cmd),
+			prometheus: parseFlagsToPrometheusConfig(cmd),
+			connection: parseFlagsToConnectionConfig(cmd),
+			reloader:   parseFlagsToReloaderConfig(cmd),
+			shipper:    parseFlagsToShipperConf(cmd),
+			tsdb:       parseFlagsToTsdbConfig(cmd),
+			objStore:   regCommonObjStoreFlags(cmd, "", false),
+		}
+
+		rl := newReloaderFromConfig(
 			log.With(logger, "component", "reloader"),
-			reloader.ReloadURLFromBase(*promURL),
-			*reloaderCfgFile,
-			*reloaderCfgOutputFile,
-			*reloaderRuleDirs,
+			conf.prometheus,
+			conf.reloader,
 		)
 
 		return runSidecar(
+			conf,
 			g,
 			logger,
 			reg,
 			tracer,
-			*grpcBindAddr,
-			time.Duration(*grpcGracePeriod),
-			*grpcCert,
-			*grpcKey,
-			*grpcClientCA,
-			*httpBindAddr,
-			time.Duration(*httpGracePeriod),
-			*promURL,
-			*promReadyTimeout,
-			*dataDir,
-			objStoreConfig,
 			rl,
-			*uploadCompacted,
-			*ignoreBlockSize,
 			component.Sidecar,
 			*minTime,
-			*connectionPoolSize,
-			*connectionPoolSizePerHost,
 		)
 	}
 }
 
 func runSidecar(
+	conf sidecarComponentConfig,
```

Comment:
I  believe it'd be better to put conf as the last argument when initializing a component. If you check the code base, you'd see injected dependencies come first.